kafka/src/main/java/org/axonframework/extensions/kafka/eventhandling/consumer/subscribable/SubscribableKafkaMessageSource.java

```diff
@@ -191,6 +191,7 @@ private void addConsumer() {
 
     private void restartOnError(RuntimeException e) {
         logger.warn("Consumer had a fatal exception, starting a new one", e);
+        fetcherRegistrations.removeIf(Registration::cancel); //retrying and keeping failed items in the list will cause memory leak
```

Comment:
Invoking the `Registration#cancel` method on all the registered Fetchers will close all of them. 
Thus, it'll also close fetchers that did not encounter any predicaments at all.

So, although you've indeed spotted a potential leak, I do not think this to be the full solution yet, @backgroundvoices.
Furthermore, adding a test case is paramount for situations like this.

Perhaps you can adjust the `restartOnError` to filter the fetcher registration that's failing?
I do think this will require adjusting the `fetcherRegistrations` list to a map, wherein we use a key for storage and retrieval. 
This will allow you to invoke `restartOnError` with the key, so that you may cancel the registration of the failing fetcher.